### PR TITLE
[None][feat] Support model_kwargs with encoder CUDA graphs

### PR DESCRIPTION
**Pull request by tingyangk**
## Dev Engineer Review

- Adds declared `model_kwargs` support for encoder CUDA graphs through `EncodeExtraInputSpec`.
- Validates names, dtypes, ranks, shapes, symbolic dimensions, and duplicate declarations.
- Supports host and device tensors with static buffers and replay-time copies.
- Undeclared tensors raise `ValueError`. Undeclared non-tensors use eager execution.
- Feature-mode and encoder-decoder encoder graphs remain unsupported.
- Review finding counts and test execution results are unavailable.

## QA Engineer Review

- `tests/unittest/llmapi/test_llm_encode.py` adds validation and replay coverage for declared inputs, invalid inputs, CPU and CUDA tensors, batch-size dimensions, eager fallback, and output parity.
- `tests/integration/defs/accuracy/test_llm_api_pytorch_encode.py` adds `test_encoder_encode_with_token_type_ids_matches_huggingface`.
- `tests/integration/test_lists/qa/llm_function_core.txt` lists the BERT Yelp integration test.
- Coverage is **sufficient** for the stated feature. The supplied evidence does not show a CI `test-db/` entry.

## Per-File QA Perspective

- `tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py`: Verify static buffer allocation, zero-filled capture, symbolic padding, and replay copies.
- `tensorrt_llm/_torch/pyexecutor/engine/runners/encoder.py`: Verify input validation, eager fallback, and correct forwarding of declared inputs.
- `tensorrt_llm/llmapi/__init__.py`: Verify `EncodeExtraInputSpec` remains publicly importable.
- `tensorrt_llm/llmapi/llm.py`: Verify documentation matches `model_kwargs` forwarding and CUDA-graph behavior.
- `tensorrt_llm/llmapi/llm_args.py`: Verify specification validation, shape and dtype resolution, symbolic-dimension rules, and duplicate-name rejection.
- `tests/unittest/llmapi/test_llm_encode.py`: Covers specification validation and CUDA-graph replay paths. No integration test-list entry is required or shown.
- `tests/integration/defs/accuracy/test_llm_api_pytorch_encode.py`: Verifies non-zero `token_type_ids` propagation and HuggingFace parity. The test is listed in the QA manual-QA list.
- `tests/integration/test_lists/qa/llm_function_core.txt`: Adds the BERT Yelp encoder test to the manual-QA list.

## Description

LLM.encode() accepts model-specific tensors via `**model_kwargs`, but the encoder runner rejected them whenever encoder CUDA graphs were enabled, so models needing an extra input had to give up graph capture.

This PR replaces that rejection with a declaration. Each tensor `kwarg` is declared in `EncodeCudaGraphConfig.extra_model_inputs`; the runner gives it a static buffer, captures every bucket with a zero-filled stand-in so the graph sees the full `forward()` signature, and copies the caller's tensor into the buffer on each replay (zero-filling the padding tail).

A spec has exactly one symbolic dimension, which is what lets it cover every bucket:

- "`num_tokens`" — scales with the packed token count (e.g. `token_type_ids`)
- "`batch_size`" — one row per request (e.g. per-request features)

Other dimensions are fixed integers. Before `forward()`, a declared `kwarg` must be a tensor matching its spec; an undeclared tensor raises (almost always a forgotten spec, and falling back silently would hide the lost speedup), while an undeclared non-tensor runs that call eagerly instead of being frozen into the graph at its capture-time value. Tensors may be on host or device.

## Example

```python
from tensorrt_llm import LLM
from tensorrt_llm.llmapi import EncodeCudaGraphConfig, EncodeExtraInputSpec

cuda_graph_config = EncodeCudaGraphConfig(
    batch_sizes=[1, 2, 4],
    num_tokens=[64, 128],
    seq_lens=[32, 64],
    enable_padding=True,
    extra_model_inputs=[
        EncodeExtraInputSpec(name="token_type_ids", shape=("num_tokens",), dtype="int32"),
        EncodeExtraInputSpec(name="per_request_feature", shape=("batch_size", 20), dtype="float32"),
    ],
)

llm = LLM(model=MODEL, encode_only=True, cuda_graph_config=cuda_graph_config)

outputs = llm.encode(
    prompts,
    token_type_ids=token_type_ids,     # (total_tokens,) int32
    per_request_feature=per_request_feats,  # (len(prompts), 20) float32
)
```

### Not supported

`extra_model_inputs` is rejected for feature-mode and encoder-decoder encoder graphs.

## Test Coverage

Unit — `tests/unittest/llmapi/test_llm_encode.py` (25 new tests):
- `TestEncodeExtraInputSpec` (11): symbolic-dim rules, reserved and non-identifier names, unknown dtypes, duplicate declarations.
- Declared `kwargs` under graphs (14): accepted; rejects undeclared / missing / wrong dtype / wrong rank / non-tensor; host and device tensors both accepted; undeclared non-tensor falls back to eager; "batch_size"-shaped inputs accepted with wrong symbolic/literal dim and dtype rejected, and graph replay matches eager.

Integration — `tests/integration/defs/accuracy/test_llm_api_pytorch_encode.py`:
- `test_encoder_encode_with_token_type_ids_matches_huggingface[bert-yelp]` — non-zero token_type_ids through the captured graph matches HuggingFace and differs from the all-zero result, proving the `kwarg` reaches the model rather than being silently dropped. Registered in `llm_function_core.txt`.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [X] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by StanleySun639 — APPROVED** [review]

File: tests/unittest/llmapi/test_llm_encode.py [hunk 1]
```diff
@@ -275,3 +277,328 @@ def test_encode_cuda_graph_and_return_raw_logits(bert_encode_llm_cuda_graph):
...
+
+
+def test_encode_batch_size_extra_replay_matches_eager(
+    bert_encode_llm, bert_encode_llm_cuda_graph_with_batch_size_extra
+):
+    """Verify batch-size static-buffer replay reproduces eager output.
+
+    BERT ignores unknown kwargs (**kwargs in forward), so the kwarg has no
+    effect on compute. The comparison directly checks that the per-spec static
+    buffer plumbing doesn't perturb the graph.
```
[the remark this answers is not in the document]
Comment (tingyangk) [reply tests/unittest/llmapi/test_llm_encode.py:595]: There's no built-in models that take `per_request_feature`. This limitation is mentioned in the test comments.

File: tests/unittest/llmapi/test_llm_encode.py [hunk 2]
```diff
@@ -275,3 +277,328 @@ def test_encode_cuda_graph_and_return_raw_logits(bert_encode_llm_cuda_graph):
...
+        assert s.symbolic_dim() == ("batch_size", 0)
+
+    def test_batch_size_symbolic_axis_nonzero(self):
+        s = EncodeExtraInputSpec(name="x", shape=(32, "batch_size"), dtype="float32")
+        assert s.resolve_shape(num_tokens=64, batch_size=8) == (32, 8)
+        assert s.symbolic_dim() == ("batch_size", 1)
+
+    def test_mixed_symbolic_dims_rejected(self):
+        """A single tensor cannot use both 'num_tokens' and 'batch_size'."""
+        with pytest.raises(Exception, match="symbolic dim"):
```
[the remark this answers is not in the document]
Comment (tingyangk) [reply tests/unittest/llmapi/test_llm_encode.py:318]: This should be trivial.

File: tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py [hunk 3]
```diff
@@ -1993,6 +2059,34 @@ def _stage_inputs(self, key: EncoderKeyType, inputs: Dict[str,
...
+            sym, axis = self._extra_input_symbolic(spec)
+            actual_len = int(user_tensor.shape[axis])
+            padded_size = self._extra_input_padded_size(spec, key)
+            if actual_len > padded_size:
+                raise ValueError(
+                    f"extra_model_inputs[{spec.name!r}]: tensor length "
+                    f"{actual_len} along symbolic {sym!r} axis exceeds "
+                    f"padded bucket size {padded_size}")
+            static_buf = static_tensors[spec.name]
+            torch.narrow(static_buf, axis, 0, actual_len).copy_(user_tensor)
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py]: In pinned mode (`prefer_pinned()` true on every non-CC GPU) `static_tensors` at :2020 is the pinned host shadow, so when the caller passes a CUDA tensor this `copy_()` is a synchronous D2H: the host blocks until all queued device work drains, then the captured H2D node at :2270 uploads the same bytes back. That is a host sync on every replay, which the `retire_staging()` event design exists to avoid, and it makes the device-tensor path (the one every new test uses) slower than the host-tensor path.

Suggested fix: don't capture the H2D for extra inputs. Leave them out of `copy_inputs()`, and in `replay()` issue an eager stream-ordered `copy_()` straight into the device static buffer before `graphs[key].replay()` (D2D for device sources, H2D from pinned for host sources). That is already what the `not self._capture_h2d_copy` branch does, so it unifies both modes. If you'd rather keep the captured-H2D design, document host tensors as the fast path and avoid CUDA tensors in the tests.

File: tensorrt_llm/_torch/pyexecutor/engine/runners/encoder.py [hunk 4]
```diff
@@ -634,13 +670,115 @@ def _reject_unsupported_model_inputs(self, model_inputs: dict[str, Any]) -> None
...
+            raise ValueError(
+                "Encoder received tensor model_kwargs not declared in "
+                f"cuda_graph_config.extra_model_inputs: {undeclared_tensor}. "
+                "Add them to the spec or disable encoder CUDA graphs. "
+                f"Declared inputs: {sorted(declared_names)}"
+            )
+
+        missing = sorted(declared_names - provided_names)
+        if missing:
+            raise ValueError(
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/pyexecutor/engine/runners/encoder.py]: Declaring a kwarg makes it mandatory on every `encode()` call. For the motivating case (`token_type_ids`) HF treats the argument as optional and defaults to zeros, and the static buffer is already zero-filled, so a missing declared input could simply zero-fill the buffer instead of raising. That would be wrong for something like `inputs_embeds`, so it's a design choice: either add an optional/required flag on `EncodeExtraInputSpec`, or keep strict and say so in the `extra_model_inputs` field description, since today a user who enables graphs for BERT must start passing `token_type_ids` on every call.

File: tensorrt_llm/_torch/pyexecutor/engine/runners/encoder.py [hunk 5]
```diff
@@ -634,13 +670,115 @@ def _reject_unsupported_model_inputs(self, model_inputs: dict[str, Any]) -> None
...
+        provided_names = set(model_inputs)
+        undeclared_names = provided_names - declared_names
+
+        undeclared_tensor = sorted(
+            name for name in undeclared_names if isinstance(model_inputs[name], torch.Tensor)
+        )
+        if undeclared_tensor:
+            raise ValueError(
+                "Encoder received tensor model_kwargs not declared in "
+                f"cuda_graph_config.extra_model_inputs: {undeclared_tensor}. "
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/pyexecutor/engine/runners/encoder.py]: This raises on an undeclared tensor even when the call could never use the graph anyway (e.g. `multi_item_part_lens` present, which `_prepare_encoder_graph_inputs` refuses at :380, or shapes outside every bucket). Not wrong, but consider checking those eager-only conditions first so multi-item scoring callers aren't forced to declare specs for a path that stays eager.

File: tensorrt_llm/llmapi/llm_args.py [hunk 6]
```diff
@@ -332,6 +461,19 @@ class EncodeCudaGraphConfig(BaseCudaGraphConfig):
         "`seq_lens` is generated from this value. Ignored by a fixed-shape "
         "feature encoder.")
 
+    extra_model_inputs: List[EncodeExtraInputSpec] = Field(
```
Comment (brnguyen2) [inline tensorrt_llm/llmapi/llm_args.py]: New field on a user-facing config: the telemetry manifest walker auto-enrolls `EncodeExtraInputSpec.shape` (tuple of `literal|int`), so `tensorrt_llm/usage/llm_args_golden_manifest.json` needs regenerating via `scripts/generate_llm_args_golden_manifest.py` and CODEOWNER approval per AGENTS.md. Also the description has an unbalanced parenthesis: `(\`num_tokens\` → ... \`max(batch_sizes)\`; pre-capture ...` never closes.

File: tests/integration/test_lists/qa/llm_function_core.txt [hunk 7]
```diff
@@ -622,6 +622,7 @@ accuracy/test_llm_api_pytorch_encode.py::TestDecoderEncode::test_decoder_encode_
 accuracy/test_llm_api_pytorch_encode.py::TestEncoderEncode::test_encoder_encode_cuda_graph_matches_eager_logits[bert-yelp]
 accuracy/test_llm_api_pytorch_encode.py::TestEncoderEncode::test_encoder_encode_matches_huggingface_classification[bert-yelp-cuda_graph]
 accuracy/test_llm_api_pytorch_encode.py::TestEncoderEncode::test_encoder_encode_matches_huggingface_classification[bert-yelp-eager]
+accuracy/test_llm_api_pytorch_encode.py::TestEncoderEncode::test_encoder_encode_with_token_type_ids_matches_huggingface[bert-yelp]
```
Comment (brnguyen2) [inline tests/integration/test_lists/qa/llm_function_core.txt]: The sibling BERT tests also run pre-merge from `tests/integration/test_lists/test-db/l0_a100.yml` (lines 32-34). Add this one there as well so graph replay with a real extra input is exercised in L0, not only in manual QA.

**Review by brnguyen2 — COMMENTED** [review]
Two things need to happen before merge, plus a few smaller items.

**Telemetry golden manifest is not regenerated.** `EncodeExtraInputSpec.shape` is `Tuple[Union[Literal[...], PositiveInt], ...]`, which the manifest walker in `tensorrt_llm/usage/llmapi_config.py` compiles as a homogeneous sequence of `literal|int`, and `_nested_models` descends into list-element models. So `cuda_graph_config.extra_model_inputs.shape` and `encoder_cuda_graph_config.extra_model_inputs.shape` should become new rows, and `tests/unittest/usage/test_llmapi_config_telemetry_docs.py` will fail against the committed JSON. Please run `python3 scripts/generate_llm_args_golden_manifest.py`, commit `tensorrt_llm/usage/llm_args_golden_manifest.json`, and get the telemetry/privacy CODEOWNER sign-off that AGENTS.md requires for new fields. (I could not run the check locally; this is from reading the walker.)

**Device tensors pay a host sync per replay** in pinned mode. See the inline comment at `[cuda_graph_runner.py:2085](https://github.com/NVIDIA/TensorRT-LLM/pull/19439/files#diff-04500eda9435bc22d845f9f29171528019dfde8a0f90a53afa51df7c4f21f76dR2085)`. Every test in this PR passes CUDA tensors, so the path users will copy from the tests is the slow one.

Smaller items:

- **Test-db entry.** `test_encoder_encode_with_token_type_ids_matches_huggingface[bert-yelp]` is added only to the QA list. Its siblings run pre-merge from `tests/integration/test_lists/test-db/l0_a100.yml:32-34`; add it there too or the replay path has no CI coverage on real weights.
- **Ticket.** This adds a public config class and changes encode() semantics under CUDA graphs; it should carry a TRTLLM JIRA rather than `[None]`.
- **Docs.** There is no user doc for encoder CUDA graphs at all (`docs/source/features/embeddings.md` does not mention them), so `extra_model_inputs` has nowhere to land. At minimum, the `EncodeCudaGraphConfig` docstring should carry a short worked example; a section in embeddings.md would be better.
- **Three hand-kept reserved-name lists** now exist for the same names: `_ENCODE_EXTRA_INPUT_RESERVED_NAMES` (llm_args.py), `_ENCODER_INPUT_KEYS_PREPARED_INTERNALLY` (encoder.py), and `_RESERVED_KEYS` in `LLM.encode()`. Consider deriving two from one.

Description matches the diff (the rendered description was truncated at "batch_siz" on my side; the code confirms both symbolic dims).
